### PR TITLE
Ast

### PR DESCRIPTION
**Pull request by jbreckman**
Need to update README.md still, but this adds a ton of functionality:
(triangles with dots in them still work)

* `list.where.(a.b.c)=5`
* `list.where.(a.b.sublist.where.(d.e)=6.0.name)=josh`
* `list.select.(a.b.sublist.select.e|(d.e))|(a.b)`
* `list.distinct.(a.b)` 
* `list.ascending.(a.b)`
* `list.stats.(a.b.c)`
* `list.group_by.(a.b.c)`

`where` is also greatly enhanced.  `sift` still works, but there are more ways to access it:

```
btoa(JSON.stringify({"name":{"$exists":true}}))
"eyJuYW1lIjp7IiRleGlzdHMiOnRydWV9fQ=="
```

* `list.where.eyJuYW1lIjp7IiRleGlzdHMiOnRydWV9fQ==` (base64 encode sift JSON)
* If the `input` to sift is a string, it will parse it

You can do more without `sift` now:

* `list.where.or(and(Country=France|Aspect=Economy)|and(Country=Australia|not(Aspect=Freedom)))`
* `list.where.Value>5`
* `list.where.Value>=5`
* `list.where.Value<5`
* `list.where.Value<=5`
* `list.where.Value!=5`

Comparison values in `where` can also be referenced by path, solved from the list object context.  It supplies a `ref` so it can also hit globals if necessary:
* `list.where.Value<=${app.data.originationHPI}`

Lastly, strings can be quoted if you have dangerous characters in them:

* `list.where.Value="This is a.... value!"` (escaping strings with `\"` works)

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,16 +1,9 @@
 var sift = require('sift'),
     clues = require('clues'),
+    pathParser = require('./build/pegjs-parser').parse,
     d3Scale = require('d3-scale'),
     Promise = clues.Promise;
 
-var reSplitter = /[\||Λ]/;
-
-// Polyfill for Object.setPrototypeOf
-Object.setPrototypeOf = Object.setPrototypeOf || function(obj, proto) {
```
Comment (jbreckman) [inline index.js:9]: i don't think this is necessary anymore...

File: index.js [hunk 2]
```diff
@@ -19,74 +12,152 @@ function setPrototype(self) {
 
 // WARNING Sift exposes access to javascript through $where
 // Here we override $where with an error
-sift.useOperator('where',function() { throw '$WHERE_NOT_ALLOWED';});
```
Comment (jbreckman) [inline index.js:22]: this was for sift 1.0.  it's up to like 12.0 now :-D

File: index.js [hunk 3]
```diff
@@ -19,74 +12,152 @@ function setPrototype(self) {
...
 // This is the main prototype 
 var Query = Object.create(Array.prototype);
 
-Query.scale = function(_domain) {
+function astToCluesPath(node) {
+  return astToString(node.paren || node);
+}
+
+function astToString(node) {
+  if (node.piped) {
```
Comment (jbreckman) [inline index.js:30]: reverse the pegjs parsing... helpful for passing to inner calls to clues

File: index.js [hunk 4]
```diff
@@ -19,74 +12,152 @@ function setPrototype(self) {
...
+    return `not(${astToString(node.not)})`;
+  }
+  if (Array.isArray(node)) {
+    return node.map(n => astToString(n)).join('.');
+  }
+  return node;
+}
+
+function createExternal(cb) {
+  let counter = 0,
```
Comment (jbreckman) [inline index.js:61]: this is most of the magic around $external working with more complex paths

File: index.js [hunk 5]
```diff
@@ -19,74 +12,152 @@ function setPrototype(self) {
...
-  return function $property(key) {
-    key = key.split(reSplitter);
-    var domainKey =  key[1] || _domain;
-    var rangeKey = key[0];
-    if (!domainKey)
+
+  return createExternal(async ast => {
+    let pipePieces = ast.piped ? ast.piped : [ast];
+    if (pipePieces.length >= 3) {
+      throw 'MAXIMUM_OF_THREE_INPUTS';
```
Comment (jbreckman) [inline index.js:87]: this is max of 2... whoops

File: index.js [hunk 6]
```diff
@@ -19,74 +12,152 @@ function setPrototype(self) {
...
+
+    let domainValues = Promise.map(self, d => clues(d, domainKey, $global).catch(noop));
+    let rangeValues = Promise.map(self, d => clues(d, rangeKey, $global).catch(noop));
+
+    domainValues = await domainValues;
+    rangeValues = await rangeValues;
+
+    // modify the items - but in a single tick, so should be safe
+    self.forEach((d,i) => {
+      d._domainValue = domainValues[i];
```
Comment (jbreckman) [inline index.js:103]: this technically wouldn't work on straight strings or numbers, but `scale` won't work for that either

File: index.js [hunk 7]
```diff
@@ -97,121 +168,195 @@ Query.scale = function(_domain) {
...
-  
-    // Provide pipe delimited filtering
-    ref = ref.split(reSplitter).sort();
-    if (ref.length > 1)
-      // Solve for the first one, and then the remainder
-      return [ref[0],function(q) {
-        return [{q:q},'q.where.'+ref.slice(1).join(reSplitter),Object];
-      }];
+function generateEvaluateConditionFn(self, ast, $global, _filters, $valueFn, pipeOperation) {
+  if (ast.piped) {
```
Comment (jbreckman) [inline index.js:178]: this is the new `where` and `where_not` support helper.  `where_not` was actually slightly different from `where` in terms of `true` and `false`

File: index.js [hunk 8]
```diff
@@ -97,121 +168,195 @@ Query.scale = function(_domain) {
...
+    return async item => {
+      let leftSide = clues(item, path, $global).catch(noop);
+      let rightSide = target;
+      
+      if (target.quoted) {
+        target = target.quoted;
+      }
+      if (target.remoteLink) {
+        try {
+          rightSide = await clues({q:self}, 'q.'+astToCluesPath(target.remoteLink), $global, 'item');
```
Comment (jbreckman) [inline index.js:215]: if i remove `'item' as a ref here, you can't access global.  I'm not sure if that's good or not.  i don't think we have secrets directly accessible in global --- and there is no way to get the results of this really.   

maybe you can brute force it?  `data.a<database_password` and then you binary search it?

File: index.js [hunk 9]
```diff
@@ -97,121 +168,195 @@ Query.scale = function(_domain) {
...
+        let p = astToCluesPath(path.equation.left);
+        return {
+          path: p,
+          key: path.equation.right
+        };
+      }
+      let p = astToCluesPath(path);
+      return {
+        path: p,
+        key: p.replace(/[.ᐉᐅ()|Λ=]/g,'_')
```
Comment (jbreckman) [inline index.js:320]: select used to only work for 1 level deep - so this flattens it out.... 

i think thats best?  not quite sure what makes most sense here

File: index.js [hunk 10]
```diff
@@ -229,52 +374,46 @@ Query.reversed = function() {
...
+      let path = astToCluesPath(ast);
+      return Promise.map(obj, d => {
+        return clues(d,path,$global).catch(noop);
+      })
+      .then(keys => {
+        obj.forEach(function(d,i) {
+          d.sortkey = $valueFn(keys[i]);
+        });
+        obj = obj.sort(function(a,b) {
+          let aNull = (a.sortkey === null || a.sortkey === undefined);
```
Comment (jbreckman) [inline index.js:394]: this is a crazy sort.  i may have written it.

File: paths.pegjs [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+StringLiteral = '"' chars:DoubleStringCharacter* '"' {
+    return {quoted:chars.join('')};  
+  }
+  
+DoubleStringCharacter
+  = '\\' '"' { return '"'; }
+  / !'"' . { return text(); }
+
+Base64FriendlyWord = [^.ᐉᐅ()<>!|Λ=\$\{\}]+[=]?[=]? {
```
Comment (jbreckman) [inline paths.pegjs:76]: i wish these were a bit cleaner....

File: index.js [hunk 12]
```diff
@@ -1,92 +1,132 @@
...
-
-// WARNING Sift exposes access to javascript through $where
-// Here we override $where with an error
-sift.useOperator('where',function() { throw '$WHERE_NOT_ALLOWED';});
-
+ 
 // Helper functions
-function toDots(d) { return d.replace(/ᐉ/g,'.'); }
 function noop() {}
+function inverseIdentity(d) { return !d; }
```
Comment (ZJONSSON) [inline index.js:15]: inverseIdentity and identity are not used here (they are in conditional.js)

File: index.js [hunk 13]
```diff
@@ -1,92 +1,132 @@
...
+  let counter = 0,
+      context = {};
+
+  return function $external(path) { 
+    let ast = pathParser(path);
+    let firstNode = ast.root;
+    let firstNodeKey = `_counter${astToString(firstNode)}`;
+    let remainder = ast.extra ? `.${ast.extra}` : '';
+
+    let existingCounter = context[firstNodeKey];
```
Comment (ZJONSSON) [inline index.js:39]: Might add a comment briefly explaining why we move the formula to `externalKey${existingCounter}`

File: index.js (hunk 8 above)
Comment (ZJONSSON) [reply index.js]: Yeah ideally nothing with possible user input should be called with a "caller" implying an internal function

File: ast.js [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+  if (Array.isArray(node)) {
+    return node.map(n => astToString(n)).join('.');
+  }
+  return node;
+}
+
+module.exports = {
+  astToCluesPath,
+  astToString,
+  pathParser: pathParser = require('./build/pegjs-parser').parse
```
Comment (ZJONSSON) [inline ast.js:51]: `pathParser =` looks redundant

**Pull request merged**